core/impl/src/main/java/com/blazebit/persistence/impl/AbstractCTECriteriaBuilder.java

```diff
@@ -201,38 +208,83 @@ public CTEInfo createCTEInfo() {
-                    PathExpression pathExpression = (PathExpression) selectExpression;
-                    // Only append the id if it's not already there
-                    if (!idAttribute.getName().equals(pathExpression.getExpressions().get(pathExpression.getExpressions().size() - 1).toString())) {
-                        pathExpression.getExpressions().add(new PropertyExpression(idAttribute.getName()));
+                    boolean firstBinding = true;
+                    final Collection<String> embeddedPropertyNames;
+
+                    if (lastAttribute.getPersistentAttributeType() == Attribute.PersistentAttributeType.EMBEDDED) {
+                        embeddedPropertyNames = new TreeSet<>(JpaMetamodelUtils.getEmbeddedPropertyNames((EmbeddableType<?>) ((SingularAttribute<?, ?>) lastAttribute).getType()));
+                    } else {
+                        embeddedPropertyNames = new TreeSet<>(cbf.getJpaProvider().getIdentifierOrUniqueKeyEmbeddedPropertyNames(cteType, attributeName));
+                    }
+
+                    PathExpression baseExpression = embeddedPropertyNames.size() > 1 ?
+                            ((PathExpression) selectExpression).clone(false) : ((PathExpression) selectExpression);
+
+                    for (String embeddedPropertyName : embeddedPropertyNames) {
+                        PathExpression pathExpression = firstBinding ?
+                                ((PathExpression) selectExpression) : baseExpression.clone(false);
+
+                        // TODO: this is supposed to check whether someAssociation is bound to someAlias.id
+                        // Actually, the id names can collide in the case of IdClass or EmbeddedIds, so a true check
+                        // should check whether the type of the path reference of the path expression is a basic type.
+                        // However, this seems not possible at this time because for some tests pathExpression.getPathReference()
+                        // returns null.
+                        if (embeddedPropertyName.equals(pathExpression.getExpressions().get(pathExpression.getExpressions().size() - 1).toString())) {
+                            throw new IllegalStateException("An association should be bound to its association type and not its identifier type");
+                        }
+
+                        pathExpression.getExpressions().add(new PropertyExpression(embeddedPropertyName));
+                        String nestedAttributePath = attributeName + "." + embeddedPropertyName;
+                        ExtendedAttribute<?, ?> nestedAttributeEntry = attributeEntries.get(nestedAttributePath);
+
+                        // Process the nested attribute path recursively
+                        attributeQueue.add(nestedAttributePath);
+
+                        // Replace this binding in the binding map, additional selects need an updated index
+                        bindingMap.remove(attributeName);
```

Comment:
Move this out of the loop

Reply:
I'm worried about the case where the embeddedProps list may be empty, in which case we do not optimize anything yet remove it from the attribute map. Could such a thing occur? I.e. if the inverse property is not mapped in an entity?

Reply:
I'm not sure this is possible. You shouldn't be able to bind an inverse property